src/macaron/parsers/bashparser.py

```diff
@@ -115,78 +148,95 @@ def parse(bash_content: str, macaron_path: str = "") -> dict:
-    repo_path : str
+    name: str
+        A name to be used as the identifier of the node.
+    node_type: BashScriptType
+        The type of the node.
+    source_path: str
+        The file that contains the bash script.
+    parsed_obj: dict | None
+        The parsed bash scrit object.
+    repo_path: str
         The path to the target repo.
-    working_dir : str
-        The working directory from which the bash script has run.
-        Empty value is considered as the root of the repo.
-    job_name: str
-        The name of the job where commands were called.
-    step_name: str
-        The name of the step where commands were called.
-
-    Yields
+    caller: BaseNode
+        The caller node.
+    recursion_depth: int
+        The number of times this function is called recursively.
+    macaron_path=None
+        The path to the Macaron module.
+
+    Returns
+    -------
+    BashNode
+        A bash node object.
+
+    Raises
     ------
-    BashCommands
-        The parsed bash script objects.
+    CallGraphError
+        When unable to create a bash node.
     """
-    if not macaron_path:
-        macaron_path = global_config.macaron_path
-
-    parsed_parent = parse(bash_content)
-    caller_commands = parsed_parent.get("commands", [])
-    if caller_commands:
-        yield BashCommands(
-            caller_path=ci_file,
-            CI_path=ci_file,
-            CI_type=ci_type,
-            commands=caller_commands,
-            job_name=job_name,
-            step_name=step_name,
-        )
+    if recursion_depth > defaults.getint("bashparser", "recursion_depth", fallback=3):
+        raise CallGraphError(f"The analysis has reached maximum recursion depth {recursion_depth} at {source_path}.")
+    parsed_content = {}
+    working_dir = None
+    match node_type:
+        case BashScriptType.INLINE:
+            if parsed_obj is None:
+                raise CallGraphError(f"Unable to find the parsed AST for the CI step at {source_path}.")
+            step_exec = validate_step(parsed_obj)
+            if step_exec is None:
+                raise CallGraphError(f"Unable to validate parsed AST for the CI step at {source_path}.")
+
+            working_dir = step_exec.get("WorkingDirectory")
+            run_script = validate_run_step(parsed_obj)
+            if run_script is None:
+                raise CallGraphError(f"Invalid run step at {source_path}.")
+            parsed_content = parse(run_script, macaron_path=macaron_path)
+        case BashScriptType.FILE:
+            parsed_content = parse_file(source_path, macaron_path=macaron_path)
+    bash_node = BashNode(
+        name, node_type, source_path, parsed_step_obj=parsed_obj, parsed_bash_obj=parsed_content, caller=caller
+    )
+    caller_commands = parsed_content.get("commands", [])
 
     # Parse the bash script files called from the current script.
-    if recursive and repo_path:
+    if caller_commands and repo_path:
```

Comment:
From the function signature, the type of `repo_path` is str, is there any case where `repo_path` is passed as an empty string so that we need to check for it here :thinking: ?

Reply:
Yes, `repo_path` can be empty in the unit tests.